### ISSUE TITLE
Text order in Quick input method not same as Windows

### ISSUE THREAD

**Issue by dollars0427** [Cangjians/ibus-cangjie#77]
For example, The order of 「秘」（竹竹）in Windows should be 4, but in ibus-cangjie is 5.
Not all the texts have this issues, such as 「神」（戈中） is fine.

Now I only find this issues on this texts:
1. 秘(竹竹)
2. 程(竹土)
3. 搜(手水)

**Comment by bochecha** [comment]
Thanks for reporting this issue! 😃 

One thing you didn't mention is the version of the Quick input method you are using. Is it version 3 or 5?

**Comment by dollars0427** [comment]
Thanks for your reply!
I tried version 3 and version 5,  but it had same problem. ☹️

**Comment by bochecha** [comment]
Alright, in that case what I'd like is for you to give me the complete list of candidates you get on Windows, **in the order you get them**, both for version 3 and 5.

Unfortunately I don't have access to a Windows machine, so I can't test this on my own. 😞 

Can you do that for the 3 codes you found? (竹竹, 竹土 and 手水)

**Comment by dollars0427** [comment]
No problem. 
I am using MAC, but OpenVanilla (https://github.com/openvanilla/openvanilla/) , input kits which I am using is support quick input and the text order actually same with Windows.  

Here are complete text lists of these 3 codes:

## 竹竹：
![2016-12-31 7 52 46](https://cloud.githubusercontent.com/assets/2662715/21577348/ccd75b2e-cf92-11e6-8705-767054af58d8.png)
![2016-12-31 7 52 50](https://cloud.githubusercontent.com/assets/2662715/21577349/ccd9c4c2-cf92-11e6-9b42-1c1f87e6e2fe.png)
![2016-12-31 7 52 54](https://cloud.githubusercontent.com/assets/2662715/21577351/ccdbc5e2-cf92-11e6-9721-15862047110b.png)
![2016-12-31 7 52 57](https://cloud.githubusercontent.com/assets/2662715/21577350/ccda6b3e-cf92-11e6-8ea4-7921249324d3.png)
![2016-12-31 7 53 01](https://cloud.githubusercontent.com/assets/2662715/21577352/ccdde764-cf92-11e6-9972-8a75fe07bd48.png)
![2016-12-31 7 53 05](https://cloud.githubusercontent.com/assets/2662715/21577353/cce0a44a-cf92-11e6-892b-0a23b2f69ddd.png)
![2016-12-31 7 53 09](https://cloud.githubusercontent.com/assets/2662715/21577354/ccfac244-cf92-11e6-8c60-d93f87b12b22.png)
![2016-12-31 7 53 13](https://cloud.githubusercontent.com/assets/2662715/21577355/ccff7c94-cf92-11e6-9897-54018f99fee3.png)

----------

## 竹土:
![2016-12-31 7 55 51](https://cloud.githubusercontent.com/assets/2662715/21577364/400ac478-cf93-11e6-9113-18c8a3d953a5.png)
![2016-12-31 7 55 55](https://cloud.githubusercontent.com/assets/2662715/21577366/400d7542-cf93-11e6-85cf-5614ad30888d.png)
![2016-12-31 7 55 59](https://cloud.githubusercontent.com/assets/2662715/21577365/400d37bc-cf93-11e6-81f0-89f158f22240.png)
![2016-12-31 7 56 03](https://cloud.githubusercontent.com/assets/2662715/21577367/401113b4-cf93-11e6-9a55-4e34c0126606.png)
![2016-12-31 7 56 08](https://cloud.githubusercontent.com/assets/2662715/21577368/4014ac72-cf93-11e6-99fb-cd2862bc78cc.png)
![2016-12-31 7 56 12](https://cloud.githubusercontent.com/assets/2662715/21577369/4014b1ea-cf93-11e6-9d5f-5986f1a0d78b.png)
![2016-12-31 7 56 17](https://cloud.githubusercontent.com/assets/2662715/21577370/402f1742-cf93-11e6-9a10-3136307aaf12.png)
![2016-12-31 7 56 21](https://cloud.githubusercontent.com/assets/2662715/21577371/40329840-cf93-11e6-83e8-6ea5371256de.png)
![2016-12-31 7 56 24](https://cloud.githubusercontent.com/assets/2662715/21577372/40331c84-cf93-11e6-94e3-490b8c20dc58.png)
![2016-12-31 7 56 28](https://cloud.githubusercontent.com/assets/2662715/21577373/40369e68-cf93-11e6-8de0-a43cea1e47e0.png)
![2016-12-31 7 56 32](https://cloud.githubusercontent.com/assets/2662715/21577374/403b6eb6-cf93-11e6-87af-a9595d9a0c8a.png)

-----------
##手水
![2016-12-31 8 01 31](https://cloud.githubusercontent.com/assets/2662715/21577397/fcfe8d44-cf93-11e6-92d4-664991eb0fcb.png)
![2016-12-31 8 01 36](https://cloud.githubusercontent.com/assets/2662715/21577396/fcfe852e-cf93-11e6-84c2-08a889514513.png)
![2016-12-31 8 01 39](https://cloud.githubusercontent.com/assets/2662715/21577398/fd01b0be-cf93-11e6-8896-a63de076feb9.png)
![2016-12-31 8 01 43](https://cloud.githubusercontent.com/assets/2662715/21577401/fd0a4292-cf93-11e6-80e6-926a7fa99559.png)
![2016-12-31 8 01 47](https://cloud.githubusercontent.com/assets/2662715/21577399/fd07acbc-cf93-11e6-8547-a87107fe6982.png)
![2016-12-31 8 01 51](https://cloud.githubusercontent.com/assets/2662715/21577400/fd08ff04-cf93-11e6-957e-67b2885b6eaf.png)

**Comment by bochecha** [comment]
(note that screenshots make it harder for me to make any comparison, it would have been easier if you had typed them)

---

Ok, so this is what libcangjie returns for `竹竹` (`h*h`) in Cangjie 3:

```
秀 身 物 秒 秘 第 穆 穋 穢 颺 釁 舠 牬 秎 秭 笏 牻 笫 稊 颩 颬 馝 篎 穇 篸 鼢 簜
```

And according to your screenshots, on OpenVanilla that would be:

```
   身 物 秒 秘 第 穆 穋 穢 颺 釁 舠 牬 秎 秭 笏 牻 笫 稊 颩 颬 馝 篎 穇 篸 鼢 簜 飂
```

*Note: I kept only the first 27 characters above (i.e 3 pages) because after that there are more differences, probably due to the default filtering in libcangjie, and those should be very rare anyway.*

From the above, it is quite clear that what happens is that we have the first character, 秀, which just doesn't seem to be anywhere in your screenshots on OpenVanilla.

Thus I have two questions:

1. Are we absolutely sure that OpenVanilla does the exact same thing as Windows?
2. Is the code for 秀 incorrect?

In our data, we have two codes for 秀: `竹木弓竹尸` (`hdnhs`) and `竹木弓尸竹` (`hdnsh`). Maybe that second one is incorrect and it should just be removed?

At this point I need @yookoala's help here. 😃 

---

Now on Cangjie 5, this is what libcangjie returns for `竹竹` (`h*h`) :

```
身 物 秒 秘 第 穆 穋 穢 颺 釁 舠 牬 秎 秭 笏 牻 笫 稊 颩 颬 馝 篎 穇 篸 鼢 簜 飂
```

And according to your screenshots, on OpenVanilla that would be:

```
身 物 秒 秘 第 穆 穋 穢 颺 釁 舠 牬 秎 秭 笏 牻 笫 稊 颩 颬 馝 篎 穇 篸 鼢 簜 飂
```

It seems to me that our Cangjie 5 results are the same as the ones you have on OpenVanilla.

So, hum... Maybe the order you're expecting is the Cangjie 5 one, and there's no problem with our Cangjie 3 ordering?

**Comment by yookoala** [comment]
On Windows 10, input "竹竹" returns these:
<img width="132" alt="windows-hh" src="https://cloud.githubusercontent.com/assets/91274/21584595/336143b8-d0e8-11e6-9ea7-8e22d3057549.png">

Seems OpenVanilla is using a different order than Windows.

**Comment by yookoala** [comment]
Fact first. These are the output for "hh" from Windows 10 Quick:

<img width="132" alt="windows-hh" src="https://cloud.githubusercontent.com/assets/91274/21584635/de0c91c2-d0e9-11e6-8557-2be05f31de9c.png">
<img width="126" alt="windows-hh-2" src="https://cloud.githubusercontent.com/assets/91274/21584636/e0856de8-d0e9-11e6-929a-fae1b9cc5e6d.png">
<img width="126" alt="windows-hh-3" src="https://cloud.githubusercontent.com/assets/91274/21584637/eadcf69e-d0e9-11e6-8621-76ece12e5029.png">
<img width="127" alt="windows-hh-4" src="https://cloud.githubusercontent.com/assets/91274/21584638/eae5e8e4-d0e9-11e6-8d63-95152a4b1338.png">
<img width="128" alt="windows-hh-5" src="https://cloud.githubusercontent.com/assets/91274/21584639/ed4f4972-d0e9-11e6-9d9a-788cd5ad9720.png">

So basically the character order is:
```
身 物 秒 秘 第 穆 穋 穢 颺 釁 舠 牬 秎 秭 笏 牻 笫 稊 颩 颬 馝 篎 穇 篸 鼢 簜 飂 飋 剓 彡 徏 毝 毶 犙 馚 鬽 彻 笌 箉 簓 躮 飏 鹄
```

Much similar to OpenVanilla's order.

From my readings:

1. Windows 10 is still using Cangjie 3 for character code and order.
2. People use OpenVanilla on Mac specifically for the Cangjie 3 code and ordering that it provides (Mac stock Cangjie is Cangjie 5)

I think we do have a problem with our Quick ordering.

**Comment by yookoala** [comment]
I've spent some times in tracing the character "秀". It is no natural for a character to have 2 sets of code. It is however mapped to 2 code in Cangjie 3 (`竹木弓竹尸` / `hdnhs` and `竹木弓尸竹` / `hdnsh`). The mapping was there all the way since 8585a29, which is when table.txt was first created.

Same case for "垂" (竹土) and "拔" (手水).

If you search for standard Cangjie code for these words, you'd find that these are non-standard mapping. I remember we added some code to match "standard character" of different governments. This issue seems to be the direct result of these mapping.

Take "秀" as an example. By standard mapping, "秀" only maps to `竹木弓竹尸` / `hdnhs`, not `竹木弓尸竹` / `hdnsh`. Quick shows characters match the first and last code in Cangjie. Therefore "秀" should only appear on Quick by `竹尸` / `hs` and not `竹竹` / `hh`.

**Comment by yookoala** [comment]
Proposed solution:

## A. Have all these duplicated mapping removed.

We can trace down the list of Cangjie 3 mapping with a "`,`" in Cangjians/libcangjie@8585a29 table.txt. Then we can simply remove the second mapping.

Pros: Simple and Clean :-)
Cons: This will be a drawback for Cangjie side.

## B. Change the frequency of the secondary mapping

Quick list character by frequency. We may simply change the frequency of any secondary mapping to 0. All the unofficial Cangjie mapping will "sink" to the bottom of the Quick list, together with some rarely used characters.

Pros: Not too difficult to do, also.
Cons: Cangjie users who looks for "unofficial" mapping would be less happy.

## My thought

I think B is in general a better solution. Strangely, I have memories that we did something similar before. Is this some sort of regression issue?

**Comment by bochecha** [comment]
> I think B is in general a better solution. Strangely, I have memories that we did something similar before. Is this some sort of regression issue?

No, it is a different one.

You remember correctly that we have already done this in the past (your proposed solution B), however that was for 氐 (see the two lines for it in `table.txt`)

This might indeed be the right solution for these 3 new cases.

However, if we end up having more and more of these, then maybe we'll need to start thinking about having different codes for Cangjie and Quick? Probably not something to answer now, but to keep in mind for the future...

**Comment by yookoala** [comment]
I think this is mainly an issue of expected order, not the code mapping. With B, Quick users will have their expected order and the non-default mapping come after. If we simply do different coding on Cangjie and Quick, then Quick users will lost the non-default mapping.

I think we should consider this "secondary frequency" as a general solution instead of special case handle. That is to change the compile process instead of the table itself.

Anyway, I search through table.txt in Cangjians/libcangjie@8585a29. After removing the x-disambiguation mappings, there are 254 occurrence of non-standard mapping in Cangjie 3 (includes `氐`, `秀`, `垂`, `拔`):

```
乃 nhs,nsh
乍 hs,os
乓 omi,omo
乖 hjlmp,hjlp
乜 pn,ps
亮 yrbn,yrbu
仍 onhs,onsh
伢 omlh,omvh
低 ohpm,ohvi
作 ohs,oos
侯 omok,onmk
修 olhh,oloh
候 olmk,olnk
偏 ohsb,oisb
傲 ogsk,oqsk
儿 hu,lu
兔 nahi,nui
兜 hshu,hvhu
兢 hujru,jujru
冢 bmmo,gbmo
冤 bnai,bnui
凄 imjln,imjlv
凳 nomrn,nomtn
函 neu,nue
匾 shsb,sisb
卑 hhj,hwkj
卵 hhsli,hhsy
卸 omsl,ovsl
及 nhe,nno
叟 hxe,hxle
后 hmr,hovie
吸 rnhe,rnno
呈 rhg,rmg
咋 rhs,ros
咎 hor,hyr
咸 ihmr,ihrm
哥 mrdr,mrnr
唬 ryphn,rypu
唳 rhsk,risk
啤 rhhj,rhwj
喂 ramv,rwmv
善 tgtr,ttr
喉 romk,ronk
嘛 ridd,rijc
圾 gnhe,gnno
坻 ghpm,ghvi
垂 hjtg,hjtm
埂 gmlk,mgmlk
墟 gypc,gypm
墨 wfg,wgfg
壹 gbmm,gbmt
奏 qkhk,qkmk
奶 vnhs,vnsh
孕 nhnd,nsnd
孵 hhbnd,hhsld
寅 jmlc,jmwc
寨 jtcd,jtce
尕 nhf,nsf
岈 umlh,umvh
岌 unhe,unno
巷 tcru,tcsu
底 ihpm,ihvi
延 nkhym,nkhyv
异 rut,sut
彝 vmfft,vmfqt
彬 ddhh,ddhhh
後 hmr,hovie
御 hooml,hoovl
微 hounk,houuk
忝 hkp,mkp
怍 phs,pos
怎 hsp,osp
憎 pcea,pcwa
戽 hsyj,isyj
戾 hsik,isik
房 hsyhs,isyhs
扃 hsbr,isbr
扇 hssmm,issmm
扈 hsrau,israu
扉 hslmy,islmy
扔 qnhs,qnsh
抵 qhpm,qhvi
押 qwj,qwl
拐 qrks,qrsh
拔 qike,qikk
捏 qag,qhxm
捩 qhsk,qisk
捶 qhjg,qhjm
掮 qhsb,qisb
搐 qyiw,qyvw
敏 owok,oyok
敖 gsok,qsok
敦 ydk,ydok
斌 ykipm,ykmpm
昨 ahs,aos
晷 ahor,ahyr
望 ybhg,ybmg
极 dnhe,dnno
柞 dhs,dos
柢 dhpm,dhvi
梁 ecd,eid
椽 dnio,dvno
榷 dlbg,dobg
殛 mnmem,mnnem
毅 yhhne,yohne
氐 hpm,hvpi
氖 onnhs,onnsh
氯 onlme,onvne
汆 oe,ohe
汲 enhe,enno
沉 ebhn,ebhu
沛 ejb,eylb
沿 ecr,ehnr
浙 eqhl,eqhml
涎 enkm,enkv
涵 eneu,enue
添 ehkp,emkp
渭 eab,ewb
漱 edlk,edlo
潢 etmc,etwc
炭 ukf,umf
炸 fhs,fos
焉 mmysf,mylf
煮 jaf,jkaf
煸 fhsb,fisb
煽 fhsm,fism
熏 hgf,hjwf
熬 gkf,qkf
牌 llhhj,llhwj
牖 llhsb,llisb
犀 sehq,syyq
犁 hdhq,hnhq
猴 khomk,khonk
獒 gkik,qkik
珊 mgbbm,mgbt
璜 mgtmc,mgtwc
瓢 mfhho,mfhvo
瓮 cimn,cimvn
甍 twlbn,twln
疆 ngmwm,nmmwm
痄 khs,kos
痹 kwml,kwt
瘟 kabt,kwot
皋 hadm,hakj
盈 nhbt,nsbt
睾 htgtj,hwgti
瞢 twbbu,twlu
瞰 bumjk,bunjk
砑 mrmlh,mrmvh
砟 mrhs,mros
砥 mrhpm,mrhvi
确 mrnbg,mrnbq
碌 mrlme,mrvne
碑 mrhhj,mrhwj
碥 mrhsb,mrisb
碳 mrukf,mrumf
磺 mrtmc,mrtwc
祆 ifhk,ifmk
祓 ifike,ifikk
祗 ifhpm,ifhvi
祚 ifhs,ifos
离 yukb,yulb
秀 hdnhs,hdnsh
秫 hdid,hdijc
程 hdrhg,hdrmg
稗 hdhhj,hdhwj
稽 hdika,hdiua
窄 jchs,jcos
窗 jchwk,jchwn
竿 hmj,tmj
笈 hnhe,hnno
笮 hhs,hos
篇 hhsb,hisb
篷 hihj,hyhj
簧 htmc,htwc
粱 ecfd,eifd
羝 tqhpm,tqhvi
羰 tqukf,tqumf
羹 tfmqc,tgftk
翁 cism,cismm
考 jkss,jkys
聚 seooo,sepho
肇 hklq,iklq
肩 hsb,isb
胙 bhs,bos
胝 bhpm,bhvi
脾 bhhj,bhwj
膳 btgr,bttr
臻 mgqkd,mgqke
舄 hxpf,hxyf
舔 hrhkp,hrmkp
舜 bbniq,bbnq
舴 hyhs,hyos
船 hycr,hyhnr
艿 tnhs,tnsh
芨 tnhe,tnno
茇 tike,tikk
莒 trhr,trr
莽 tikt,tkt
菝 tqie,tqik
菡 tneu,tnue
蓄 tyiw,tyvw
蓑 tyrv,tywv
蔑 twlh,twli
蘸 tmgf,tmwf
虎 yphn,yphu
虞 yprmk,yprvk
虮 lihn,livii
蚕 hklmi,mklmi
蚱 lihs,lios
蜒 linkm,linkv
蝙 lihsb,liisb
融 mblmi,mrlmi
螯 gklmi,qklmi
蟥 litmc,litwc
蠢 qalii,qkali
衰 yrmv,ywmv
衷 ykhv,ylhv
袍 fprux,lpru
袒 ifam,lam
袖 iflw,llw
袼 lher,yher
褊 lhsb,lisb
角 nbg,nbq
起 goru,gosu
趴 rmc,rmho
趿 rmnhe,rmnno
跋 rmike,rmikk
跚 rmbbm,rmbt
蹁 rmhsb,rmisb
辣 yjdb,yjdl
迓 ymlh,ymvh
迮 yhs,yos
述 yid,yijc
透 yhdh,yhds
逞 yrhg,yrmg
逸 ynai,ynui
遍 yhsb,yisb
邸 hinl,hmnl
酢 mwhs,mwos
釜 cc,ckmgc
阼 nlhs,nlos
隙 nlfaf,nlfhf
雇 hsog,isog
雌 ympog,ypog
非 lmsy,lmyyy
面 mwsl,mwyl
饔 ygoiv,yvgv
魃 hiike,hiikk
魏 hdhi,hvhi
麻 idd,ijcc
黻 lbike,tbikk
鼎 buux,buvml
鼐 nhbux,nsbul
```

**Comment by yookoala** [comment]
Any update on this issue?

**Comment by bochecha** [comment]
> Any update on this issue?

I'm not sure what kind of update you're expecting. I'm fine with implementing your option B above for those 3 characters we identified in this issue, as I had said in https://github.com/Cangjians/ibus-cangjie/issues/77#issuecomment-269959217

As for the more general solution, let's move the discussion to a separate issue.

**Comment by bochecha** [comment]
@dollars0427 Hi there! We didn't forget you and this issue, it only turned out to be harder to fix than we initially thought. Sorry for the wait. :slightly_smiling_face: 

Are you comfortable with the command line? I'd like to send you a file for you to test, but that will require entering a few commands as root.

**Comment by dollars0427** [comment]
@bochecha Yes, I can. There are a computer which installed Ubuntu 14.04 so I would allow to using command line as root.

Thanks for your help !

**Comment by bochecha** [comment]
@dollars0427 Alright, so ideally I'd like you to do this on the machine where you use ibus-cangjie normally. Do the following:

1.  download a test libcangjie database:
    ```
    $ wget https://www.daitauha.fr/static/cangjie.db
    ```
2.  backup the current libcangjie database:
    ```
    $ sudo cp /usr/share/libcangjie/cangjie.db /usr/share/libcangjie/cangjie.db.bak
    ```
3.  install the test libcangjie database:
    ```
    $ sudo cp cangjie.db /usr/share/libcangjie/cangjie.db
    ```
4. reboot the computer;

After a reboot, I'd like you to use ibus-cangjie as usual, and pay attention to two things:

1. the character ordering you reported above, are they fixed? (that is, the order is now the same as on Windows for them)
2. do you see **new** ordering problems, which didn't exist before?

I'd like you to use ibus-cangjie with this test database for some time (maybe a week or so?), to get a better impression of whether I fixed the problem, or whether I introduced new issues.

Thanks! :slightly_smiling_face: 

---

If this creates too many problems for you, you can easily rollback the changes:

```
$ sudo cp /usr/share/libcangjie/cangjie.db.bak /usr/share/libcangjie/cangjie.db
```

Then reboot, and everything should be back as it was, including the present bug.

**Comment by dollars0427** [comment]
Thank you for your help! I tested the new database on my computer, it's look like fixed many order problem. It's great!

However, I find that there are new ordering problem. For example,
The ordering of 「水弓」is:
![show_reel16](https://user-images.githubusercontent.com/2662715/39974620-96ea82fe-575c-11e8-85cb-ac6b42dc18c6.png)

The ordering on Hime (It's same as Windows, too.) is:
![show_reel16](https://user-images.githubusercontent.com/2662715/39974738-34a5145a-575d-11e8-94d3-de16ab153fe7.png)

 I will attach the ordering on Windows later. 

Thank you !

**Comment by bochecha** [comment]
I'd like to ask you the following, if possible, as they make it much easier for me to work on this issue:

1. instead of screenshots, the list of characters **as text** allows me to copy-paste;
2. when you give a code, 「水弓」 is useful, but saying that it corresponds to 'en' (the keys on the US keyboard) makes it easier to identify;

The reason for these 2 is that I can't read Chinese, and I can't type Cangjie either. It might sound surprising, but it's the truth.

I do manage to figure things out from your report, but it takes me a lot of time, and the 2 points above greatly simplify investigations for me. :slightly_smiling_face: 

In addition, are we 100% certain that Hime has the same order as Windows for all possible inputs? In general, please give directly the Windows order.

Having said that…

---

For 「水弓」 (`en`), your screenshots show that : 

* 汀沉汽沆沸洌涔涮湔 with libcangjie
* 汀　汽沆沸洌涔涮湔 with Hime

So we can clearly see that we added 沉 in there, moving everything else back.

沉 has 2 codes, `ebhn` and `ebhu`. Can you give me the order on Windows for 「水山」 (`eu`)?

**Comment by bochecha** [comment]
Hi @dollars0427!

Sorry for the lack of reply for so long. We didn't forget you, but we've been very busy with other things and couldn't work on Cangjie much. :slightly_frowning_face: 

Since February I have been gathering all the results for all the codes on Windows. It took ages, but I finally got them all. (the gory details are at https://gitlab.freedesktop.org/cangjie/platform-comparisons/ if you're interested)

That allowed me to compare with the results from libcangjie and… we're much worse than I thought. :disappointed: 

However, it also allowed me to work on a couple of systematic fixes which dramatically improve the compatibility without much effort.

Like last time, I'd like to ask you to test a database file, and tell me if you can still find issues.

I'll write again the instructions below so you don't need to search for them in the previous comments.

I'd like you to do this on the machine where you use ibus-cangjie normally. Do the following:

1.  download a test libcangjie database:
    ```
    $ wget https://www.daitauha.fr/static/cangjie-110.db
    ```
2.  backup the current libcangjie database:
    ```
    $ sudo cp /usr/share/libcangjie/cangjie.db /usr/share/libcangjie/cangjie.db.bak
    ```
3.  install the test libcangjie database:
    ```
    $ sudo cp cangjie-110.db /usr/share/libcangjie/cangjie.db
    ```
4. reboot the computer;

After a reboot, I'd like you to use ibus-cangjie as usual, and pay attention to two things:

1. the character ordering you reported above, are they fixed? (that is, the order is now the same as on Windows for them)
2. do you see **new** ordering problems, which didn't exist before?

I'd like you to use ibus-cangjie with this test database for some time (maybe a week or so?), to get a better impression of whether I fixed the problem, or whether I introduced new issues.

Thanks! :slightly_smiling_face: 

---

If this creates too many problems for you, you can easily rollback the changes:

```
$ sudo cp /usr/share/libcangjie/cangjie.db.bak /usr/share/libcangjie/cangjie.db
```

Then reboot, and everything should be back as it was, including the present bug.

**Comment by dollars0427** [comment]
Thank you very much.
I am sorry that I always forget to check the github notification,  so reply too late  ;)
I will try this fix on my Ubuntu 16.04 computer and report the result to you later.

**Comment by bochecha** [comment]
> I am sorry that I always forget to check the github notification, so reply too late ;)

It's never too late. Especially not after we gave you no response for over a year. :slightly_smiling_face:

**Comment by peteroid** [comment]
Hi @bochecha, thanks for sharing the instructions in https://github.com/Cangjians/ibus-cangjie/issues/77#issuecomment-509034968 and I confirmed that the ordering works as expected. Thanks a lot! :tada: :tada: :tada:
